File: mops/benchmarks/sasaw.cpp
#include <algorithm>
#include <iostream>
#include <vector>

#include "mops.hpp"
#include "utils.hpp"

int main() {

    auto output = std::vector<double>(1000 * 100 * 32);
    fill_vector_random_floats(output);

    auto A = std::vector<double>(60000 * 13);
    fill_vector_random_floats(A);

    auto B = std::vector<double>(60000 * 32);
    fill_vector_random_floats(B);

    auto C = std::vector<double>(900);
    fill_vector_random_floats(C);

    auto W = std::vector<double>(1000 * 7 * 32);
    fill_vector_random_floats(A);

    auto indices_A = std::vector<int32_t>(900);
    fill_vector_random_integers(indices_A, 13);

    auto indices_W_1 = std::vector<int32_t>(60000);
    fill_vector_random_integers(indices_W_1, 1000);

    auto indices_W_2 = std::vector<int32_t>(900);
    fill_vector_random_integers(indices_W_2, 7);

    auto indices_output_1 = std::vector<int32_t>(60000);
    fill_vector_random_integers(indices_output_1, 1000);

    auto indices_output_2 = std::vector<int32_t>(900);
    fill_vector_random_integers(indices_output_2, 100);

    for (int i = 0; i < 100; i++) {
        mops::sparse_accumulation_scatter_add_with_weights<double>(
            {output.data(), {1000, 100, 32}},
            {A.data(), {60000, 13}},
            {B.data(), {60000, 32}},
            {C.data(), {900}},
            {W.data(), {1000, 7, 32}},
            {indices_A.data(), {900}},
            {indices_W_1.data(), {60000}},
            {indices_W_2.data(), {900}},
            {indices_output_1.data(), {60000}},
            {indices_output_2.data(), {900}}
        );
    }

    return 0;
}

File: python/mops-torch/tests/sasaw.py
import mops.torch
import pytest
import torch
from mops.reference_implementations import (
    sparse_accumulation_scatter_add_with_weights as ref_sasaw,
)
from mops.torch.reference_implementations import (
    sparse_accumulation_scatter_add_with_weights as ref_sasaw_torch,
)

torch.manual_seed(0xDEADBEEF)


def test_sasaw():
    A = torch.rand(100, 20)
    B = torch.rand(100, 200)
    W = torch.rand(25, 13, 200)
    C = torch.rand(50)
    indices_output_1 = torch.randint(25, size=(100,), dtype=torch.int32)
    indices_W_1 = torch.randint(25, size=(100,), dtype=torch.int32)
    output_size = 15
    indices_A = torch.randint(20, size=(50,), dtype=torch.int32)
    indices_W_2 = torch.randint(13, size=(50,), dtype=torch.int32)
    indices_output_2 = torch.randint(output_size, size=(50,), dtype=torch.int32)

    reference = torch.tensor(
        ref_sasaw(
            A.numpy(),
            B.numpy(),
            C.numpy(),
            W.numpy(),
            indices_A.numpy(),
            indices_W_1.numpy(),
            indices_W_2.numpy(),
            indices_output_1.numpy(),
            indices_output_2.numpy(),
            output_size,
        )
    )
    actual = mops.torch.sparse_accumulation_scatter_add_with_weights(
        A,
        B,
        C,
        W,
        indices_A,
        indices_W_1,
        indices_W_2,
        indices_output_1,
        indices_output_2,
        output_size,
    )
    assert torch.allclose(reference, actual)


@pytest.mark.parametrize("dtype", [torch.float64])
@pytest.mark.parametrize("device", ["cpu", "cuda"])
def test_sasaw_grads(dtype, device):
    A = torch.rand(70, 10, dtype=dtype, device=device, requires_grad=True)
    B = torch.rand(70, 50, dtype=dtype, device=device, requires_grad=True)
    W = torch.rand(3, 5, 50, dtype=dtype, device=device, requires_grad=True)
    C = torch.rand(50, dtype=dtype, device=device)
    indices_output_1 = torch.randint(3, size=(70,), dtype=torch.int32, device=device)
    indices_W_1 = torch.randint(3, size=(70,), dtype=torch.int32, device=device)
    output_size_2 = 15
    indices_A = torch.randint(10, size=(50,), dtype=torch.int32, device=device)
    indices_W_2 = torch.randint(5, size=(50,), dtype=torch.int32, device=device)
    indices_output_2 = torch.randint(
        output_size_2, size=(50,), dtype=torch.int32, device=device
    )

    if device != "cuda":  # not yet implemented
        assert torch.autograd.gradcheck(
            mops.torch.sparse_accumulation_scatter_add_with_weights,
            (
                A,
                B,
                C,
                W,
                indices_A,
                indices_W_1,
                indices_W_2,
                indices_output_1,
                indices_output_2,
                output_size_2,
            ),
        )

    if device != "cuda":  # not yet implemented
        assert torch.autograd.gradgradcheck(
            mops.torch.sparse_accumulation_scatter_add_with_weights,
            (
                A,
                B,
                C,
                W,
                indices_A,
                indices_W_1,
                indices_W_2,
                indices_output_1,
                indices_output_2,
                output_size_2,
            ),
        )


def test_sasaw_ref():
    A = torch.rand(70, 10)
    B = torch.rand(70, 50)
    W = torch.rand(3, 5, 50)
    C = torch.rand(50)
    indices_output_1 = torch.randint(3, size=(70,), dtype=torch.int32)
    indices_W_1 = torch.randint(3, size=(70,), dtype=torch.int32)
    output_size_2 = 15
    indices_A = torch.randint(10, size=(50,), dtype=torch.int32)
    indices_W_2 = torch.randint(5, size=(50,), dtype=torch.int32)
    indices_output_2 = torch.randint(output_size_2, size=(50,), dtype=torch.int32)

    reference = torch.tensor(
        ref_sasaw(
            A.numpy(),
            B.numpy(),
            C.numpy(),
            W.numpy(),
            indices_A.numpy(),
            indices_W_1.numpy(),
            indices_W_2.numpy(),
            indices_output_1.numpy(),
            indices_output_2.numpy(),
            output_size_2,
        )
    )
    actual = ref_sasaw_torch(
        A,
        B,
        C,
        W,
        indices_A,
        indices_W_1,
        indices_W_2,
        indices_output_1,
        indices_output_2,
        output_size_2,
    )
    assert torch.allclose(reference, actual)
